vpr/src/noc/sat_routing.cpp

```diff
@@ -0,0 +1,986 @@
+        }
+    } else {
+        agg_bw_expr = 0;
+    }
+
+
+    orsat::LinearExpr congested_link_sum = orsat::LinearExpr::Sum(congested_link_vars);
+    congested_link_sum *= congestion_weight;
+
+    orsat::LinearExpr objective = latency_overrun_sum + agg_bw_expr + congested_link_sum;
+    return objective;
+}
+
+
+vtr::vector<NocTrafficFlowId, std::vector<NocLinkId>> noc_sat_route(bool minimize_aggregate_bandwidth,
+                                                                    const t_noc_opts& noc_opts,
+                                                                    int seed) {
+    vtr::ScopedStartFinishTimer timer("NoC SAT Routing");
+
+    // Used to add variables and constraints to a CP-SAT model
+    orsat::CpModelBuilder cp_model;
+
+    /* For each traffic flow and NoC link pair, we create a boolean variable.
+     * When a variable associated with traffic flow t and NoC link l is set,
+     * it means that t is routed through l.*/
+    t_flow_link_var_map flow_link_vars;
+
+    /* A boolean variable is associated with each NoC link to indicate
+     * whether it is congested.*/
+    vtr::vector<NocLinkId, orsat::BoolVar> link_congested_vars;
+
+    /* Each traffic flow latency constraint is translated to how many NoC links
+     * the traffic flow can traverse without violating the constraint.
+     * These integer variables specify the number of additional links traversed
+     * beyond the maximum allowed number of links.
+     */
+    std::map<NocTrafficFlowId, orsat::IntVar> latency_overrun_vars;
+
+    create_flow_link_vars(cp_model, flow_link_vars, latency_overrun_vars);
+
+    constrain_latency_overrun_vars(cp_model, flow_link_vars, latency_overrun_vars);
+
+    forbid_illegal_turns(flow_link_vars, cp_model);
+
+    create_congested_link_vars(link_congested_vars, flow_link_vars, cp_model, noc_opts.noc_sat_routing_bandwidth_resolution);
+
+    add_continuity_constraints(flow_link_vars, cp_model);
+
+    auto objective = create_objective(cp_model, flow_link_vars, latency_overrun_vars, link_congested_vars,
+                                      noc_opts.noc_sat_routing_bandwidth_resolution,
+                                      noc_opts.noc_sat_routing_latency_overrun_weighting,
+                                      noc_opts.noc_sat_routing_congestion_weighting,
+                                      minimize_aggregate_bandwidth);
+
+    cp_model.Minimize(objective);
+
+    orsat::SatParameters sat_params;
+    if (noc_opts.noc_sat_routing_num_workers > 0) {
+        sat_params.set_num_workers(noc_opts.noc_sat_routing_num_workers);
+    }
+    sat_params.set_random_seed(seed);
+    sat_params.set_log_search_progress(noc_opts.noc_sat_routing_log_search_progress);
+
+    orsat::Model model;
+    model.Add(NewSatParameters(sat_params));
+
+    orsat::CpSolverResponse response = orsat::SolveCpModel(cp_model.Build(), &model);
+
+    if (response.status() == orsat::CpSolverStatus::FEASIBLE ||
```

Comment:
I'd write some status messages saying what happens here if it isn't already printed by the SAT solver. The messages should be pretty user-friendly / high-level, unless the caller already prints those.